From ac8d81787476f7802ca1a826371f8ff651348ef9 Mon Sep 17 00:00:00 2001
From: Riccardo Mottola <rm@gnu.org>
Date: Thu, 27 Aug 2026 17:11:12 +0200
Subject: [PATCH] Bug 1336344 - use sizeof(pthread_mutex_t) to size MutexImpl's
 platformData_; r=fitzgen

For pthreads platforms, we have a terribly large condition for the size
of a MutexImpl that attempts to hardcode the number of words that
pthread_mutex_t takes.  This hardcoding isn't always correct.  We
originally did this to try and keep <pthread.h> includes out of the
headers, but the PlatformConditionVariable.h header already includes
<pthread.h> anyway, and <pthread.h> isn't so namespace-invasive as
<windows.h>.  So go ahead and include <pthread.h> and use the actual
definition of pthread_mutex_t to size the platformData_ member.
---
 mozglue/misc/PlatformMutex.h | 22 ++++++++++------------
 1 file changed, 10 insertions(+), 12 deletions(-)

diff --git a/mozglue/misc/PlatformMutex.h b/mozglue/misc/PlatformMutex.h
index ed275d68530..f70e4f2f83f 100644
--- a/mozglue/misc/PlatformMutex.h
+++ b/mozglue/misc/PlatformMutex.h
@@ -10,6 +10,10 @@
 #include "mozilla/Attributes.h"
 #include "mozilla/Move.h"
 
+#if !defined(XP_WIN)
+# include <pthread.h>
+#endif
+
 namespace mozilla {
 
 namespace detail {
@@ -40,19 +44,13 @@ class MutexImpl
 
   PlatformData* platformData();
 
-// Linux and maybe other platforms define the storage size of pthread_mutex_t in
-// bytes. However, we must define it as an array of void pointers to ensure
-// proper alignment.
-#if defined(__APPLE__) && defined(__MACH__) && defined(__i386__)
-  void* platformData_[11];
-#elif defined(__APPLE__) && defined(__MACH__) && defined(__amd64__)
-  void* platformData_[8];
-#elif defined(__linux__)
-  void* platformData_[40 / sizeof(void*)];
-#elif defined(_WIN32)
-  void* platformData_[6];
+#if !defined(XP_WIN)
+  void* platformData_[sizeof(pthread_mutex_t) / sizeof(void*)];
+  static_assert(sizeof(pthread_mutex_t) / sizeof(void*) != 0 &&
+                sizeof(pthread_mutex_t) % sizeof(void*) == 0,
+                "pthread_mutex_t must have pointer alignment");
 #else
-  void* platformData_[64 / sizeof(void*)];
+  void* platformData_[6];
 #endif
 
   friend class mozilla::detail::ConditionVariableImpl;
